### PR TITLE
ENH: Use Dragon4 algorithm to print floating values

### PR DESCRIPTION
**Pull request by ahaldane**
This is a C port of [Ryan Juckett's C++ implementation (link)](http://www.ryanjuckett.com/programming/printing-floating-point-numbers/part-4/) of the Dragon4 algorithm for printing floating-point values to numpy. See earlier discussion in #9919 and #9932, which are both replaced by this PR. Please check the license at the top of the dragon4.c file to confirm it is suitable.

This is a WIP and I am still deciding some of the behavior, but it is in "working" state. I am putting it up now to run tests, and if anyone wants to try it out. Here's a quick summary:

 * Adds a python interface to the Dragon4 algorithms in `np.core.multiarray.dragon4_positional` and `np.core.multiarray.dragon4_scientific`. Documentation of arguments is in the C code, but you can call it with a single python floating scalar argument to try it out.
 * Numpy scalars use the dragon4 algorithm in "unique" mode (see below) for str/repr, in a way that tries to match python float output. Note: one possible bugfix is that the scientific mode cutoff for all floats is now 1e16, instead of 1e10 for float32 and 1e17 for float64 like before. CPython's cutoff is 1e16. This also fixes complex reprs like in #9332.
 * updates `arrayprint.py` code to use dragon4. I also added a new option, that you can set `np.printoptions(precision='unique')`, which will print the minimum number of digits to uniquely specify the value. I left the default at `precision=8` for now. 
 * In `arrayprint.py` I've also stolen/reworked the `FormatFloat` generalizations from #9919. Now all floating types just use a single `FloatingFormat` formatter, which calls the dragon4 routines.

Some examples that you can compare to the old behavior:
```python
print(complex(0,np.inf), np.complex128(complex(0,np.inf)), str(np.array([np.complex128(complex(0,np.inf))])))
np.set_printoptions(precision='unique')
print(np.arange(10, dtype='f4')/10.)
a = np.random.rand(4)
print(repr(a))
print(repr(a.astype('f4')))
print(repr(a.astype('f2')))
print(repr(10.**np.arange(20)))
```

I still need to do some cleanups and tests. Also, there may be a bug in the original implementation because `np.float64(1e23)` isn't nicely rounded, unlike the other powers of 10.

### CODE REVIEW

**Comment by ahaldane** [comment]
Output of the code in my last comment, with this PR:
```
(infj, infj, '[ 0.+infj]')
[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
array([0.41337804458542926, 0.9634011951702381 , 0.38924457984371086,
       0.6889281918519422 ])
array([0.41337803, 0.9634012 , 0.3892446 , 0.6889282 ], dtype=float32)
array([0.4133, 0.9634, 0.3892, 0.689 ], dtype=float16)
array([1.e+00, 1.e+01, 1.e+02, 1.e+03, 1.e+04, 1.e+05, 1.e+06, 1.e+07,
       1.e+08, 1.e+09, 1.e+10, 1.e+11, 1.e+12, 1.e+13, 1.e+14, 1.e+15,
       1.e+16, 1.e+17, 1.e+18, 1.e+19])
```

Output with master:
```
(infj, inf*j, '[ 0.+infj]')
[ 0.          0.1         0.2         0.30000001  0.40000001  0.5
  0.60000002  0.69999999  0.80000001  0.89999998]
array([ 0.5864939 ,  0.15714977,  0.26403041,  0.7634838 ])
array([ 0.58649391,  0.15714976,  0.2640304 ,  0.76348382], dtype=float32)
array([ 0.58642578,  0.15710449,  0.26391602,  0.76367188], dtype=float16)
array([  1.00000000e+00,   1.00000000e+01,   1.00000000e+02,
         1.00000000e+03,   1.00000000e+04,   1.00000000e+05,
         1.00000000e+06,   1.00000000e+07,   1.00000000e+08,
         1.00000000e+09,   1.00000000e+10,   1.00000000e+11,
         1.00000000e+12,   1.00000000e+13,   1.00000000e+14,
         1.00000000e+15,   1.00000000e+16,   1.00000000e+17,
         1.00000000e+18,   1.00000000e+19])
```

**Comment by eric-wieser** [comment]
Out of curiosity - what keeps us from directly using the C++ version?

**Comment by eric-wieser** [comment]
```
array([0.41337803, 0.9634012 , 0.3892446 , 0.6889282 ], dtype=float32)
```
looks kinda bad to me, but I understand the rationale.

What does this give?
```
a = 10.**np.arange(20)
a[0] = 1234
```

**Comment by ahaldane** [comment]
Re: the spacing in the float32 array: That is behavior is still to be decided, now that we have more choices due to the existence of a dragon4 interface. Note that example only gets output for `precision='unique'`, but we don't need to make that the default. (In fact, it is not the default in this PR... note I manually set it in my example).

Also, here is what this PR currently outputs by default for your code:
```python
[1]: a = 10.**np.arange(20)
[2]: a[0] = 123456
[3]: a
array([1.23456e+05, 1.00000e+01, 1.00000e+02, 1.00000e+03, 1.00000e+04,
       1.00000e+05, 1.00000e+06, 1.00000e+07, 1.00000e+08, 1.00000e+09,
       1.00000e+10, 1.00000e+11, 1.00000e+12, 1.00000e+13, 1.00000e+14,
       1.00000e+15, 1.00000e+16, 1.00000e+17, 1.00000e+18, 1.00000e+19])
```

**Comment by ahaldane** [comment]
Also, another comment on the spacing in `unique` mode:

We already get funny spacing like that in current numpy master:
```python
[31]: array([0.32254803, 0.8584671, 0.07836795, 0.2622], dtype=float32)
array([ 0.32254803,  0.8584671 ,  0.07836795,  0.2622    ], dtype=float32)
```

**Comment by ahaldane** [comment]
As for leaving it in C++, I honestly didn't consider that carefully since I assumed it would be harder/more confusing that porting to C and trying to interface to it.  Also, while I didn't modify the dragon4 code itself much, I did pretty heavily rewrite the "wrapper" part of the code that converts dragon4 output to a string.

**Comment by ahaldane** [comment]
More thoughts on spacing: I don't think padding with 0s is correct, since the "unnecessary" digits which were dropped are not necessarily 0.

To avoid the trailing spaces, probably we could do as follows: In the first pass in FloatingFormat, figure out the maximum fractional length of the data output will all the trimming. Then, set the precision to that value in the second pass, and don't do any trimming in the second pass (trim-mode 'k'), thus keeping the extra digits in the values that were a bit shorter in the first pass.

**Comment by mdickinson** [comment]
Presumably this would fix #9360? (And also close #2643 and #6136.)

**Comment by charris** [comment]
The test failures are due to ieee extended precision on 32 linux being aligned on 4 byte boundaries, hence `float96` instead of `float128`.

**Comment by ahaldane** [comment]
Yes, I thought I accounted for that with the `#ifdef NPY_FLOAT96`, but apparently that macro-constant isn't defined...

**Comment by ahaldane** [comment]
Also I figured out the `1e23` strangeness: It is because the implementation currently doesn't use IEEE unbiased rounding. It should be easy to modify it to do so.

Also, in principle it is possible to change the rounding mode, see [gcc's discussion (link)[https://www.gnu.org/software/libc/manual/html_node/Rounding.html]. However, I think CPython has forced/assumed the rounding mode to be IEEE unbiased, so I think it is safe to assume the same for numpy.

File: numpy/core/src/multiarray/dragon4.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#include "numpy/arrayobject.h"
+#include "npy_config.h"
+#include "npy_pycompat.h"
+#include "numpy/arrayscalars.h"
+
+typedef enum TrimMode
+{
+    TrimMode_None,         // don't trim zeros, always leave a decimal point
```
Comment (charris) [inline numpy/core/src/multiarray/dragon4.h]: Should convert these to C style comments.

**Comment by charris** [comment]
> Yes, I thought I accounted for that with the #ifdef NPY_FLOAT96, but apparently that macro-constant isn't defined...

Missing include statement. Need `#include <numpy/npy_common.h>`, maybe something in `setup` also, depending

**Comment by mdickinson** [comment]
> CPython has forced/assumed the rounding mode to be IEEE unbiased

Assumed, rather than forced. The dtoa.c code is a bit of a mix: some parts (mainly the integer arithmetic parts) explicitly construct the round-ties-to-even results, without help from the FPU. Others (the fast paths that use floating-point), assume that the FPU control word remains at its default of round-ties-to-even. If the FPU rounding mode were set to something else, it's probable that Python's `float` repr would give wrong results in some corner cases, and conceivable that it would do something even more horrible, like entering an infinite loop. On the one hand, that's probably bad, and CPython should (but doesn't) make an effort at least to check that the FPU control word is doing the right thing. On the other hand, it doesn't _seem_ to have caused any error reports to date, and the code has been in use for a good number of years now.

So yes, while in principle assuming that the x87/SSE2 rounding mode settings will be the right ones makes me a bit nervous, in practice it appears to be a fairly safe assumption.

Also, the dtoa.c code is written in such a way that infinite loops could happen (at one point I started refactoring to eliminate the potential for infinite loops, but that project never got completed); I'd hope that the Dragon4 code is a bit better behaved in that respect.

[Full disclosure: I was responsible, with Eric Smith, for implementing the CPython shortest repr implementation, though of course the original core `dtoa.c` code comes from David Gay.]

**Comment by ahaldane** [comment]
@mdickinson thanks for the links to the issues.

Assuming IEEE unbiased rounding in this PR is "safe" in the sense that it won't crash. I think the worst case scenario is that we spit out floating point strings that don't covert back to their original values, eg in the case of 1e23. (By the way, I am discussing the rounding mode issues with Ryan Juckett over on his website, linked above).

I don't really want to go through the effort of accounting for all the rounding modes here. But at least maybe I could add a check to `fegetround` during numpy startup that gives a serious warning if the rounding mode isn't unbiased.

**Comment by ahaldane** [comment]
Updated
 * Implemented IEEE unbiased rounding
 * fixed a possible bug in the Dragon4 implementation
 * did a pass at reformatting everything to match numpy style (including C comments)

When I come back to this I'm going to work on an arrayprint.py interface and on choosing the right defaults. I'm going to try adding a`floatmode` printing option, which controls controls the interpretation of the `precision` printing option. The doc would be:
```
    floatmode : str, optional
        Controls the interpretation of the `precision` option for
        floating-point types. Can take the following values:
            - 'fixed' : Always print exactly `precision` fractional digits,
                    even if this would print more or fewer digits than
                    necessary to specify the value uniquely.
            - 'unique : Print the minimum number of fractional digits necessary
                    to represent each value uniquely. Different elements may
                    have a different number of digits.  The value of the
                    `precision` option is ignored.
            - 'maxprec' : Print at most `precision` fractional digits, but if
                    an element can be uniquely represented with fewer digits
                    only print it with that many.
            - 'maxprec_equal' : Print at most `precision` fractional digits,
                    but if every element in the array can be uniquely
                    represented with an equal number of fewer digits (or less),
                    use that many digits for all elements.
        Default is 'maxprec'
```

**Comment by ahaldane** [comment]
Whew. I just spent 2 hours to determine that the pure-Python computation `1e23 == 10.**23` is architecture-dependent. It's `True` on most of our test systems, but `False` on the 2nd, 3rd and 4th appveyor (windows) systems. Probably it's due to some detail of the OS `powf` implementation.

So if you want to generate all allowed `float64` powers of 10 (or the closest IEEE-unbiased rounded numbers), doing `10.**arange(308)` doesn't work in windows. It is safer to do `np.array(['1e{}'.format(i) for i in range(308)], dtype='f8')`. On linux the two are equivalent.

Also, windows does not have `fenv.h` so we can't easily check the rounding mode there, so to be pragmatic I will always just assume it is `FT_TONEAREST` without any tests.

**Comment by ahaldane** [comment]
All right, I think this is finished. 

This PR now fixes a number of very old float-representation problems, and also fixes up the LongFloat output as promised in time for 1.14, by generalizing the printing code for any float size. You can get an overview of the PR from the release notes, from the docstrings of `float_format_*`, and from the new tests (2nd to last commit).

After chatting with Ryan Juckett on his website I'm convinced that the core dragon4 implementation is correct, including the few small modifications that Ryan made relative to the Burger & Dybvig paper, plus some small tweaks of my own for IEEE unbiased rounding.

I've tried to arrange the commits in a way that shows which parts were just the port from C++ to C, and which parts were modifications. It might be easier to understand what was changed by going through the commits one by one instead of looking at all the changes at once. Let me know if I can reorganize it in some way that is helpful.

**Comment by charris** [comment]
> Probably it's due to some detail of the OS powf implementation.

Kudos tracking that down. Sometimes I think NumPy is a better test of math libraries than the actual library maintainers :)

**Comment by charris** [comment]
I haven't reviewed in detail, but this looks pretty clean and I suspect you have a far better understanding of its contents than I, so let's get it in. Thanks Allan.

**Comment by ahaldane** [comment]
Whew that was quick!

I'll go through and close the relevant PRs, I think I needed commas in the commit message because they were not auto-closed.

**Comment by charris** [comment]
I think you need a separate `closes` for each issue.

**Comment by charris** [comment]
The rounding still worries me and might be a problem. I suppose we need to rely on the tests to check if that is a problem. Also, out of curiosity, what were the python (hence compiler) versions that had trouble with 1e23?

File: numpy/core/tests/test_scalarprint.py [hunk 2]
```diff
@@ -5,26 +5,193 @@
...
+        assert_equal(fsci64('299792458.0', **preckwd(5)), "2.99792e+08")
+
+        assert_equal(fpos32('3.14159265358979323846', **preckwd(25)),
+                            "3.1415927410125732421875000")
+        assert_equal(fpos64('3.14159265358979323846', **preckwd(50)),
+                         "3.14159265358979311599796346854418516159057617187500")
+        assert_equal(fpos64('3.14159265358979323846'), "3.141592653589793")
+
+
+	# smallest numbers
```
Comment (eric-wieser) [inline numpy/core/tests/test_scalarprint.py]: nit: bad indentation here

File: numpy/core/arrayprint.py [hunk 3]
```diff
@@ -463,6 +483,23 @@ def array2string(a, max_line_width=None, precision=None,
         (whitespace character) in the sign position of positive values.  If
         '-', omit the sign character of positive values. If 'legacy', print a
         space for positive values except in 0d arrays.
+    floatmode : str, optional
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Are we sure we don't want `float_mode` for consistency with `suppress_small`?

File: numpy/core/arrayprint.py [hunk 4]
```diff
@@ -736,75 +937,22 @@ def __call__(self, x):
...
-        elif x >= 0:
-            if self.sign == '+':
-                return '+' + format_longfloat(x, self.precision)
-            else:
-                return ' ' + format_longfloat(x, self.precision)
-        else:
-            return format_longfloat(x, self.precision)
-
-
-class LongComplexFormat(object):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: I think we might need to keep this for compatibility, as a subclass that does nothing but throw a warning in it's constructor.

File: numpy/core/arrayprint.py [hunk 5]
```diff
@@ -597,8 +634,8 @@ def _formatArray(a, format_function, rank, max_line_len,
                           summary_insert).rstrip()+']\n'
     return s
 
-class FloatFormat(object):
-    def __init__(self, data, precision, suppress_small, sign=False):
+class FloatingFormat(object):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Same comment about leaving a public `FloatFormat` object visible, assuming we're considering this public API.

Docstring saying `For np.floating objects` or something

File: numpy/core/arrayprint.py [hunk 6]
```diff
@@ -736,75 +937,22 @@ def __call__(self, x):
         return self.truestr if x else "False"
 
 
-class LongFloatFormat(object):
-    # XXX Have to add something to determine the width to use a la FloatFormat
-    # Right now, things won't line up properly
-    def __init__(self, precision, sign=False):
+class ComplexFormat(object):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Maybe `ComplexFloatingFormat` to match `np.complexfloating`

**Review by eric-wieser — COMMENTED** [review]
Really pleased to see all the arrayprint formatters merge together, but a couple nits about names and compatibility in `arrayprint`, which we should probably look at before 1.14

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 7]
```diff
@@ -828,107 +712,107 @@ timedeltatype_str(PyObject *self)
...
- * #kind = str, repr#
- * #KIND = STR, REPR#
+ * #name = float, double, longdouble#
+ * #Name = Float, Double, LongDouble#
+ * #NAME = FLOAT, DOUBLE, LONGDOUBLE#
  */
 
-#define PREC @NAME@PREC_@KIND@
+static PyObject *
+@name@type_@kind@_either(npy_@name@ val, TrimMode trim_pos, TrimMode trim_sci,
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: Could do with a comment explaining what this function is for

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 8]
```diff
@@ -828,107 +712,107 @@ timedeltatype_str(PyObject *self)
...
-        fputs(buf, fp);
-        Py_END_ALLOW_THREADS
-        return 0;
-}
 
-#if @hascomplex@
-static int
-c@name@type_print(PyObject *v, FILE *fp, int flags)
+static PyObject *
+halftype_@kind@(PyObject *self)
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: Is this special compared to the other ones? Why?

File: numpy/core/tests/test_arrayprint.py [hunk 9]
```diff
@@ -290,29 +290,110 @@ def test_sign_spacing(self):
...
+
+        # unique mode
+        np.set_printoptions(floatmode='unique')
+        assert_equal(repr(x),
+    "array([0.6104  , 0.922   , 0.457   , 0.0906  , 0.3733  , 0.007244,\n"
+    "       0.5933  , 0.947   , 0.2383  , 0.4226  ], dtype=float16)")
+        assert_equal(repr(y),
+    "array([0.2918820979355541 , 0.5064172631089138 , 0.2848750619642916 ,\n"
+    "       0.4342965294660567 , 0.7326538397312751 , 0.3459503329096204 ,\n"
+    "       0.0862072768214508 , 0.39112753029631175])")
```
Comment (eric-wieser) [inline numpy/core/tests/test_arrayprint.py]: Indentation is pretty funky here too

File: numpy/core/arrayprint.py [hunk 10]
```diff
@@ -321,15 +347,9 @@ def _get_format_function(data, **options):
...
             return formatdict['int']()
     elif issubclass(dtypeobj, _nt.floating):
-        if issubclass(dtypeobj, _nt.longfloat):
-            return formatdict['longfloat']()
-        else:
-            return formatdict['float']()
+        return formatdict['float']()
     elif issubclass(dtypeobj, _nt.complexfloating):
-        if issubclass(dtypeobj, _nt.clongfloat):
-            return formatdict['longcomplexfloat']()
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: This means that users setting a `longcomplexfloat` formatter will now see it being ignored silently

**Comment by eric-wieser** [comment]
Deleted my comments about whitespace, and just added a STY commit to #9962, to reduce noise here

**Pull request merged**